### PR TITLE
[hail] JoinPoints for JVM backend

### PR DESCRIPTION
**Pull request by iitalics**
## `JoinPoint` interface for nontrivial, type-safe control flow in the JVM backend

This PR implements a "join-point" abstraction for the JVM backend. Join-points are a primitive
control flow construct that allow sophisticated forms of branching to be implemented in a type safe
way, without having to directly manipulate labels and jumps at the JVM bytecode level. Importantly,
they will enable the kinds of branching needed by stream-deforestation techniques that
@patrick-schultz and I have been discussing, for which while loops and if's were not sufficient.

We've also discussed plans for implementing join points as a feature in the IR.

### Notable examples:

* Implementation of `whileLoop` (emits bytecode identical to current version):
```scala
def whileLoop(cond: Code[Boolean], body: Code[Unit]): Code[Unit] =
  JoinPoint.CallCC[Unit] { (jb, break) =>
    val continue = jb.joinPoint()
    val loopBody = jb.joinPoint()
    continue.define { _ => JoinPoint.mux(cond, loopBody, break) }
    loopBody.define { _ => Code(body, continue(())) }
    continue(())
  }
```

* Mutual recursion:
```scala
def parity(
  n: Code[Int],
  even: Code[Ctrl],
  odd: Code[Ctrl]
): Code[Ctrl] = {
  val isEven = jb.joinPoint[Code[Int]](mb)
  val isOdd = jb.joinPoint[Code[Int]](mb)
  isEven.define { i => (i ceq 0).mux(even, isOdd(i - 1)) }
  isOdd.define { i => (i ceq 0).mux(odd, isEven(i - 1)) }
  isEven(n)
}
```

### Classes of interest (tl;dr)

- `JoinPoint` - Non-returning function. Used to implement control flow in a type-safe, functional way.
- `ParameterPack` - Trait used for tuple deforesting. Allows join-points to be provided multiple
  arguments.
- `JoinPointBuilder` - Used to define new join points.
- `CallCC` - Entry-point for expressions with complex control flow. Provides a `JoinPointBuilder`
  and a `JoinPoint` to return a value from the expression.

### `JoinPoint`

A `JoinPoint[A]` acts like a non-returning function with an argument of type `A`.  The type of an
applied join point is `Code[Ctrl]`, which indicates that this code does some sort of control flow
(like a jump, or a loop), instead of returning a value. Under the hood, `JoinPoint`s are implemented
with a label and a `GOTO` instruction.

```scala
def example1(j: JoinPoint[Code[Int]]): Code[Ctrl] = Code(
  j(3),
  "this line is never reached".println,
  j(4))
```

### `ParameterPack`

The trait `ParameterPack[A]` says that the type `A` is comprised of a list of `Code[T]`s, such that the structure of that list is statically known. While this is generally useful for representing deforested tuples, it is specifically useful here because it allows `JoinPoint`s to take multiple arguments as a tuple of the individual arguments. This "tuple" will be represented at runtime by pushing or popping multiple values from the JVM argument stack.

```scala
// case class JoinPoint[A: ParameterPack] ( .... )
def example2(j: JoinPoint[(Code[Int], Code[Boolean], Code[Int])]): Code[Ctrl] =
  j((8, true, 9))
  /* LDC 8
   * LDC 1
   * LDC 9
   * GOTO j */
```

### `JoinPointBuilder`

The only way to create new join-points is via a `JoinPointBuilder` object, using the `joinPoint[A]` method (where `A` is the desired argument type). Then the body of the join-point can be provided by calling `define`.

```scala
val mb: MethodBuilder
// method builder required to create locals to store join point arguments
def example3(jb: JoinPointBuilder, exit: JoinPoint[Code[Int]]): JoinPoint[Code[Int]] = {
  val j = jb.joinPoint[Code[Int]](mb)
  j.define { n => exit(n + 1) }
  j
}
```

### `CallCC`

The only way to obtain a `JoinPointBuilder` is through a `CallCC`. `CallCC` also provides access to
the "current continuation" -- a join-point which takes an argument that will become the final result
of the entire `CallCC` expression. In fact, the only way to return a value from a `CallCC` is by
eventually calling this join-point.

```scala
val example4: Code[Boolean] =
  JoinPoint.CallCC { (jb: JoinPointBuilder, ret: JoinPoint[Code[Boolean]]) =>
    const(false).mux(
      ret(false),
      ret(true))
  }
// running 'example4' gives 'true'
```

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+import is.hail.asm4s._
+import org.objectweb.asm.Opcodes
+import org.objectweb.asm.tree.{LabelNode, AbstractInsnNode, JumpInsnNode}
+import scala.collection.mutable
+import scala.collection.generic.Growable
+
+object JoinPoint {
+  abstract class CallCC[A: ParameterPack](mb: MethodBuilder) {
+    def apply[X](jb: JoinPointBuilder[X], ret: JoinPoint[A, X]): Code[X]
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala]: Ah, that makes sense, thanks. This does seem like a design flaw in the current system, but your workaround is good for now.

**Comment by iitalics** [comment]
just pushed some changes to remove the need for all the polymorphism

**Comment by patrick-schultz** [comment]
Nice! I'll have a real look on Monday, but this looks just about ready to merge.

File: hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    *   }
+    *
+    * The above code typechecks in Scala, but at runtime would corrupt the JVM stack, leaving a "1"
+    * behind without consuming it. This stack-indicator machinery catches this before bytecode
+    * verification.
+    */
+
+  class EmitLongJumpError extends AssertionError("cannot jump out of nested CallCC")
+
+  private val stack: mutable.Stack[Object] = new mutable.Stack
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala]: Using objects as the scope identifiers is clever, but it's unnecessary allocation. Why not just have a `private var scopeID: Int = 0` which is incremented on each `CallCC`?

File: hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      Code(p.push(args), gotoLabel)
+    }
+
+  private[joinpoint] def placeLabel = Code(label)
+  private[joinpoint] def gotoLabel = Code[Ctrl](new JumpInsnNode(Opcodes.GOTO, label))
+}
+
+class DefinableJoinPoint[A: ParameterPack] private[joinpoint](
+  args: ParameterStore[A],
+  _stackIndicator: Object
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala]: I don't think we ever use the underscore naming convention in scala, which has actual access control.

File: hail/src/test/scala/is/hail/asm4s/JoinPointSuite.scala [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    odd: Code[Ctrl]
+  ): Code[Ctrl] = {
+    val isEven = jb.joinPoint[Code[Int]](mb)
+    val isOdd = jb.joinPoint[Code[Int]](mb)
+    isEven.define { i => (i ceq 0).mux(even, isOdd(i - 1)) }
+    isOdd.define { i => (i ceq 0).mux(odd, isEven(i - 1)) }
+    isEven(n)
+  }
+
+  def whileLoop(cond: Code[Boolean], code: Code[Unit]*): Code[Unit] =
```
Comment (patrick-schultz) [inline hail/src/test/scala/is/hail/asm4s/JoinPointSuite.scala]: Could we make this the actual implementation of `Code.whileLoop`?

File: hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    def emit(il: Growable[AbstractInsnNode]): Unit = {
+      stack.push(si)
+      try
+        c.emit(il)
+      finally
+        assert(stack.pop() eq si)
+    }
+  }
+
+  private[joinpoint] def ensureStackIndicator[T](si: Object)(c: Code[T]) = new Code[T] {
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala]: Whoops, sorry, the `JoinPoint` code should be
```scala
def apply(args: A): Code[Ctrl] = {
  ensureStackIndicator(stackIndicator)
  Code(p.push(argss), gotoLabel)
}
...
```

[the remark this answers is not in the document]
Comment (iitalics) [reply hail/src/main/scala/is/hail/asm4s/joinpoint/JoinPoint.scala]: :+1: made this change

**Comment by iitalics** [comment]
added some logic to allow `CallCC` code to be emitted more than once in the same function (see `testDuplicateCallCC`, which used to cause bytecode verification errors).

**Pull request merged**